doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,90 @@
+..
+    Copyright (c) 2004-2015 Ned Batchelder
+    SPDX-License-Identifier: MIT
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+
+.. _codegen_intro:
+
+Introduction
+############
+
+Python snippets that are inlined in a source file are used as code generators.
+The tool to scan the source file for the Python snippets and process them is
+Codegen. Codegen and part of this documentation is based on
+`Cog <https://nedbatchelder.com/code/cog/index.html>`_ from Ned Batchelder.
+
+The processing of source files is controlled by the CMake extension functions:
+zephyr_sources_codegen(..) or zephyr_library_sources_codegen(..). The generated
+source files are added to the Zephyr sources. During build the source files are
+processed by Codegen and the generated source files are written to the CMake
+binary directory.
+
+The inlined Python snippets can contain any Python code, they are regular
```

Comment:
Does this have security implications with ANY Python code allowed?  I suppose thorough code reviews would catch something malicious being added, and I suspect it's hard to limit the allowed Python snippets.